src/agents/smart_agent/smart_agent.py

```diff
@@ -42,15 +50,31 @@ def run(self, user_input: str | None, conversation=None, stream=False) -> AgentR
         if user_input is None:  # if no input return init message
             return AgentResponse(conversation=self._conversation, response=self._conversation[1]["content"])
         
-        if conversation is not None:  # if no history return init message
+        if conversation is not None:
             self._conversation = conversation
 
         execution_error_count = 0
         run_count = 0
+
         self._conversation.append({"role": "user", "content": user_input})
+        self.__history.clean_up_history(max_q_with_detail_hist=self.__max_question_with_detail_hist, max_q_to_keep=self.__max_question_to_keep)
 
         while True:
             response_message: ChatCompletionMessage
+            if run_count >= self.__max_run_per_question:
```

Comment:
this code looks duplicated? I think it should be re-merged with line 78 to use your history object.